router/odp.go

```diff
@@ -1,25 +1,57 @@
 package router
 
 import (
+	"fmt"
+	"net"
+	"syscall"
+
 	"github.com/weaveworks/go-odp/odp"
 )
 
 // ODP admin functionality
 
-func CreateDatapath(dpname string) error {
+func CreateDatapath(dpname string) (err error, supported bool) {
 	dpif, err := odp.NewDpif()
 	if err != nil {
-		return err
+		if odp.IsKernelLacksODPError(err) {
+			return nil, false
+		}
+
+		return err, true
 	}
 
 	defer dpif.Close()
 
-	_, err = dpif.CreateDatapath(dpname)
+	dp, err := dpif.CreateDatapath(dpname)
 	if err != nil && !odp.IsDatapathNameAlreadyExistsError(err) {
-		return err
+		return err, true
+	}
+
+	// Pick an ephemeral port number to use in probing for vxlan
+	// support.
+	udpconn, err := net.ListenUDP("udp4", nil)
+	if err != nil {
+		return err, true
```

Comment:
It's significant.  The boolean determines how errors are treated. There are four possibilities:

error|supported|meaning
-------|-------------|------------
nil     | true       | ODP supported, everything fine
!= nil | true      | there was an error, don't really know whether ODP is supported or not
nil     | false    | ODP not supported (so use a normal bridge)
!= nil | false   | ODP not supported (so use a normal bridge), and the error describes why

The (error, bool) encoding seemed like the natural one in go, but I see how it isn't obvious (particularly given the tendency for an error return value to mean that other return values should be ignored - I'd hoped to signal that is not the case here with the unusual ordering of the return values).